### ISSUE TITLE
ETag validity on composite resource

### ISSUE THREAD

**Issue by Dragomir-Ivanov** [rs/rest-layer#157]
Lets suppose we have this request:
```
GET http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55?fields=*,services,visits
If-None-Match: "11b22a28b07ddf482f8269050e9691d9"
```
If I execute this in a row, first time I get 200 then 304, which is as it is supposed to be.
But then I change some field values of `services` and `visits` behind the scenes.

Then retry the query above, I still get 304, although the result of the query IS CHANGED.

ETag comparison has its place on updating concrete resources, but I think we should leave generation of ETags for compound resources, and for resouseLists to be generated in a middleware function executed after `rest-layer`.

**Comment by rs** [comment]
Do you mean `*,services{*},visits{*}`? Because without `{}` you should just get the object ids.

This is a very valid point. What about computing a composite ETag?

cc @smyrman

**Comment by Dragomir-Ivanov** [comment]
@rs `services` and `visits` are bound to `clients` by `schema.Reference` and not `schema.Connection` so `*,services,visits` results in an arrays of `services:[]` and `visits:[]` attached to the returned clients item.

Composite ETag can't be cached in the DB, so we need to recompute it on the fly, compare with the supplied ETag and return 304 if they match. 
Is all this work related to `rest-layer` at all, aren't we putting too many responsibilities on it? Other frameworks will do this in a middle-ware function executed after the frameworks.

I guess if we have some context ( but not Go context.Context) exported in the request, and we put some important values there to let the following middle-ware functions make decisions based on them.
I am against `context.Context` because it is hierarchical, and adding values in a child `ctx` doesn't propagate to parent `ctx`.

**Comment by Dragomir-Ivanov** [comment]
Okay, I think now I am understanding why middle-ware path is not considered here. In Go `http.ResponseWriter` is single shot, and once you `Write()` there is no turning back. That is why `Alice` middle-ware can be used only for modifying the `http.Request` object, but not the response.
Please confirm that my understanding is correct.

**Comment by rs** [comment]
Yes. That’s why you can set a custom ResponseSender on `rest.Handler`.

Computing a checksum of all the Etags shouldn’t be too expensive.

**Comment by Dragomir-Ivanov** [comment]
@rs Thanks, then custom` ResponseSender` is the way to go.

> Computing a checksum of all the Etags shouldn’t be too expensive.
What do you mean by this? I thought the ETags are check-sums themselves. My thinking is computing ETag for composite and list resources will be the right way, and this might be done in the `ResponseSender` mentioned above. However we shouldn't enforce that, but maybe put an example for this particular case in the examples.

**Comment by rs** [comment]
Etag is an opac string. We can combine all etags of the (sub-)resources using the hash algo of our choosing to create a composite etag.

Why do you want to fix this issue outside of rest layer? It is a real problem that need to be addressed.

**Comment by Dragomir-Ivanov** [comment]
Well, now that it is clear we can't make post `rest-layer` middle-ware function to modify the response, we have no choice. I wanted because I have seen some Node frameworks doing in middle-ware functions, and it seemed self-contained problem. 

Now, why would you went to combine all ETags? Wouldn't be the same if you only calculate ETag (md5 hash) to the composite resource response? Combining multiple-ETags seems more work.

**Comment by rs** [comment]
Summing n string is less expensive than summing n maps with arbitrary number of fields (and subfields) stored as `interface{}`. We could checksum the JSON representation, it would be in-between in term of complexity but it would also be too late and tie the Etag computation to the representation which is handled by the `rest` package instead of `resource` package.

Functionally, I agree, both approaches are equivalent.

**Comment by rs** [comment]
Just looked at the code. Accessing the Etag during projection evaluation might be tricky actually. I think `query.Projection.Eval` should return a list of resources used during the evaluation for instance, so the caller can do additional work. I means that Etag comparison won't be able to happen before projection evaluation.

We have the same problem with `If-Modified-Since` and `Last-Modified` actually.

**Comment by Dragomir-Ivanov** [comment]
What summing algorithm do you propose for Etags? Append all Etags and then do final md5 on the whole byte array, or XORing them all?

**Comment by rs** [comment]
I would go for md5. The xor route would be interesting but would require to either guarantee all Etags are same length or use padding. All rest-layer Etags are currently same size but it could change in the future or we may want to support pre-existing Etags with different format.

**Comment by smyrman** [comment]
Will the md5 hashes Etag be set in the header, while the raw MD5s are made available as `_etag` in each resource? Will the original Etag be available as a different header? How does the system respond if a resource is mussung an E-tag, e.g. as discussed in rs/rest-layer#20.
2
One -case to be aware of:
1. A user fetches a resource with some other resource embedded `GET http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55?fields=*,services,visit`
2. A user try to update the root resource if it has not changed `POST http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55 If-Match X`

**Comment by rs** [comment]
On solution would be to have a composite Etag with a separator like `<root resource etag>-<composite projection etags>`. The full Etag would be used by browser for caching, but we could strip the second part when the item is edited.

**Comment by Dragomir-Ivanov** [comment]
I think it is best if each resource brings its own `_etag` field just like in the `itemList` case.
Then we will have ETags to update freely the `root resource` or each of the `projected resources`.
I guess we will need to supply `_updated` as well.
Also when PUT/POST-ing a resource, we can look for `If-Match` HTTP header, if non we can fallback to trying to find `_etag` field within the resource, and trying to use that. This will not help browser caching, but will prevent concurrent resource update, which is more important.

**Comment by rs** [comment]
I’m not a fan of poluting the payload with metadata.

**Comment by Dragomir-Ivanov** [comment]
Okay, but we are polluting the payload in `itemList` case, aren't we? Adding `_etag` in the `referenced resources` as well would service the same purpose. Pollution of every kind is not wanted, but can't we make this configurable, or better supplying enough of the data, so custom `ResponseFormatter` will do the job.

**Comment by rs** [comment]
What’s wrong with the composite etag described above?

**Comment by Dragomir-Ivanov** [comment]
There is one particular use-case, where you get `root resource`, and a `referenced resource` item/items. Then you want to PUT/PATCH a `referenced resource` item. In your proposal, we can't do that, because we don't have the ETag for this individual item, but just for the `root resource` and  for all `projected resource` items Etag combined sum. 
If we want to update a `referenced resource` item then we will need to GET it again, then PUT. Not needing doing this is nice optimization, and will be in line of `itemList` case, where we have array of items, each with `_etag` field. In the `root resource` we also have array of `referenced resources` items, but without `_etag` field..

I understand that all this is optional, so maybe the best place to do this formatting is in the `ResponseFormatter` so the user can overwrite it.

**Comment by smyrman** [comment]
I suppose we have three or four cases for the composite e-tag:
- list view : only composite etag ("root item?")
- list view with embeds : another composite etag
- detail view : only root item etag
- detail view with embeds : root item and conposite e-tag

I suppose this will let you update the root-resource without a data race, but not an embedded sub-resource, e.g. if you want to fetch all resources through one big request with embeds, and update individual resources later without a re-fetch.

> I’m not a fan of poluting the payload with metadata.

Neither am I. However, as @Dragomir-Ivanov points out, the `rest` package _already_ include an "_etag" field in some cases. Why not do so cinsistently, and let that field be used for updates, and the header value E-tag for GET?

**Comment by smyrman** [comment]
Btw, making the header values e-tag of format `<root resource etag>-<composite projection etags>` when apllicable is still nice in order to make it less confusing to the user why the elE-tag in the header and payload sometimes do not match

**Comment by Dragomir-Ivanov** [comment]
@rs Any thoughts on "polluting" the sub-resource payload with `_etag`?

**Comment by rs** [comment]
I have no better solution to propose so let’s do that :/

**Comment by Dragomir-Ivanov** [comment]
I was tinkering with the `Etag` for the composite resource, and it turned out that during the `resource` and `sub-resource` projection evaluation different parts of `sub-resource` projections can come from DB at different times, because some sub-resources are obtained with go-routines. This is a problem because for the same request, sometimes sub-resources can come in A,B,C order, but for other times they can come in B,C,A order, thus using `md5.Hash()` function producting different composite resource hasg/ETag, for exactly the same result. @rs Can you rethink your decision on using XOR of all ETags instead. It will produce the same result no matter the order.

**Comment by smyrman** [comment]
> Can you rethink your decision on using XOR of all ETags instead.

I suppose that means that rs/rest-layer-mongo#20 would need to change to a compatible format?

> sometimes sub-resources can come in A,B,C order, but for other times they can come in B,C,A order,

Have you actually observed this? From me just skimming the code, it looks like the order is pretty much guaranteed (unless they come in a random order from the Storer layer).

Expanded references are inserted at a given map field location, so evaluation order should not matter.
- https://github.com/rs/rest-layer/blob/2c4c847a7e4d13e1245c93990665aaeb1a35cfe2/schema/query/projection_evaluator.go#L95

Connections are inserted at a given slice index, and aren't fetched concurrently as such:
- https://github.com/rs/rest-layer/blob/2c4c847a7e4d13e1245c93990665aaeb1a35cfe2/schema/query/projection_evaluator.go#L122

**Comment by smyrman** [comment]
I suppose it _is_ true that they can come from the DB at different times, but does that matter?

**Comment by Dragomir-Ivanov** [comment]
Maybe I am not understanding the code enough, but for composite resource's ETag we will need to hash with MD5: root resource ETag, sub-resource A ETag, sub-resource B ETag, etc. We will need to apply the MD5 has in the same order, for us to have the same composite-resource ETag. We will need ether store&compute this hash after all sub-resources have been obtained from storage, or use algorithm where ordering doesn't matter (XOR).
@smyrman  The snippets above seem to be executed by separate go-routines, so there is no guarantees about their scheduling, and which finishes when.
I may be wrong, but that is my understanding for now.

**Comment by smyrman** [comment]
> We will need ether store&compute this hash after all sub-resources have been obtained from storage, or use algorithm where ordering doesn't matter (XOR).

Ok, I follow you now. I was assuming you would do the first, compute the hash after all sub-resources have been fetched. If you want to do the incremental hash _while_ you are fetching resources, then yes, you would need to use another algorithm than MD5.

**Comment by smyrman** [comment]
> From me just skimming the code, it looks like the order is pretty much guaranteed (unless they come in a random order from the Storer layer).

I was thinking about the order being consistent in the final JSON response, not in terms of evaluation order, which is what you ask for -- sorry for the confusion.

**Comment by rs** [comment]
In the light of this, xor is the way to go. We can handle different sizes using padding when necessary.

**Comment by smyrman** [comment]
> We can handle different sizes using padding when necessary.

That only works if the non-conforment ID has a length _shorter_ then the typical MD5 sub. Maybe better to just do a MD5-sum of any non-conferment IDs (not == 32) befoere the XOR?

**Comment by smyrman** [comment]
Also, a final control questions before we dive into XOR.

1. What is the colission rate for using XOR for hash combinations. Is it significant?
2. XOR mapps identical hashes to zero, e.g. XOR(a,a) and XOR(b,b) both result in zero. Presumably this may be an issue for an object that potentially maps to the same object twice: `{"author": "a", "commiter": "a"}`, `{"author": "b", "commiter": "b"}`.  Most likely though, it would still be a case of `XOR(c,a,a) != XOR(d,b,b)`
3. As the order is lost in XOR, is there any plausible situation where the order in the response actually matters, so that an unordered algorithm becomes inconsistent per _unique_ URL including query parameters? I can't think of any myself.

**Comment by Dragomir-Ivanov** [comment]
Here is some discussion on the topic:
https://crypto.stackexchange.com/questions/1368/is-it-a-good-idea-to-use-bitwise-xor-on-a-set-of-md5-sums

I guess collecting all the `_etag` fileds, sort them then do the final MD5 sum is another way of doing it.

**Comment by smyrman** [comment]
@Dragomir-Ivanov , I found the same link I was reading the same one Seams the answer to 1 is `2.938735877055719e-39`. That is probably good enough.

> I guess collecting all the _etag fileds, sort them then do the final MD5 sum is another way of doing it.

If we still don't care about point 3, then yes.

However, I think XOR will work nicely. I just want to ask some control questions in case there is anything we haven't thought about:-)

**Comment by Dragomir-Ivanov** [comment]
On point 2: `XOR(c,a,a) == XOR(c)` Don't know how big is the chance of occurrence in `rest-layer`, but there is potential for it.

**Comment by smyrman** [comment]
That is a very good point. I suppose it means that for the previous example of`{"author": "a", "commiter": "a"}`,

The E-tag from `GET /commits?limit=1` would match the case where we embed them `GET /commits?limit=1,fields=comitter{*},author{*}` (or as the syntax is now, `GET /commits?limit=1,fields=comitter,author`)

**Comment by smyrman** [comment]
Can you think of any other cases where `XOR(c,a,a) == XOR(c)` could be an issue?

**Comment by smyrman** [comment]
I suppose this is actually a real case:
1. `GET /commits?limit=1,fields=comitter,author` E-tag is generated as `XOR(c,a,a)`
2. We change the object linked to by committer and author
3. We do another . `GET /commits?limit=1,fields=comitter,author` E-tag is generated as `XOR(c,b,b)`, and BOOM.

If you then set limt from `1` to `1000`, there is actually a pretty high potential error rate for the E-tag generation via XOR, as any change to the linked objects here are ignored.

**Comment by Dragomir-Ivanov** [comment]
Difference of the above options:
(1) Will ignore ordering of appearance in the `sub-resources`. For example, having sub resource field `"commits": [ c1, c2, c3]` on a different run, the DB can return these commits in a different order say: `[c2,c3,c1]`, but since the individual `_etags` will be the same, option (1) will produce the same composite ETag. Option (2) will compute different ETag for each slightest change in the ordering.

There is a problem with adding `_etag` to each sub-resource item: If we add it, then a change in the resource outside current projection ( say we project `?fields=a,b` , but something else changes `c` ), we will have different `_etag` so the composite ETag will be different, although the requested fields are going to be the same. I think we should make adding sub-resource `_etag` optional via HTTP header maybe. If we had the `*` field selection we could do it like `?fields=*,_etag`. 

Guys these corner cases start to become quite complex, so I will understand if you want to drop the ball. I guess I can live with re-fetch&update.

**Comment by smyrman** [comment]
> There is a problem with adding _etag to each sub-resource item: If we add it, then a change in the resource outside current projection ( say we project ?fields=a,b , but something else changes c ), 

This only matters for MD5 hash of the full JSON payload though, as the semantic weak E-tags generated by a hash of an alphanumerically sorted list of E-tags would change no matter if `_etag` is included in the response or not.

Besides that, I like the idea of making the `_etag` field optional to fetch by making it hidden by default. This also would also resolve @rs concern about adding meta-data to the response body. That's probably handled as a separate issue/PR?

I am leaning a bit against staring out with strong e-tags if the performance is not significantly reduced by this.  I found some benchmarks here: http://webcache.googleusercontent.com/search?q=cache:hqoueUAiJNcJ:blog.madewithdrew.com/post/benchmark-md5/+&cd=1&hl=no&ct=clnk&gl=no&client=safari

We might want to do some quick benchmarks ourselves. If it's really less than 1ms extra for a 500k response, that doesn't sound to alarming.

**Comment by rs** [comment]
If we do strong etags we will lose the ability to make conditional write requests. Or we need to always keep the root resource stored etag as a prefix to be used for writes.

**Comment by Dragomir-Ivanov** [comment]
@smyrman Currently there is no way of obtaining `Hidden` fields. If you include them in the `?fields=` parameter rest-layer returns an error. However I see a benefit of having "hidden" fields forcibly show via query, so I guess we can relax this, and maybe add another type of `Forbidden` field type, where it will behave like current `Hidden` field type.

I guess there are a lot of corner cases to cover. For example if someone choose not to include `_etag` in resource items ( root/sub ) then we can freely switch from MD5 of all resource `_etag`  to MD5 to full JSON. 
In doing so we can calculate the same ETag for modified root/sub, if that modification is not evident in the composite resource projection.
Doing so, we will not be able to use `ETag` GET HTTP response header for root resource modification, since it will be fingerprint for the whole response payload. If one wants to update single item be it the `root` one, or one of the `sub-resources`, it will need to use the `_etag` field from the item itself.
I understand this is tough decision to make, so don't rush. I am sure we will make the right decision in the end.

**Comment by smyrman** [comment]
> If we do strong etags we will lose the ability to make conditional write requests. Or we need to always keep the root resource stored etag as a prefix to be used for writes.

@rs, we will loose that anyway for the compound E-tag, either it's a strong E-tag or a hash of hashes.

**Comment by smyrman** [comment]
> @smyrman Currently there is no way of obtaining Hidden fields. If you include them in the ?fields= parameter rest-layer returns an error. However I see a benefit of having "hidden" fields forcibly show via query, so I guess we can relax this, and maybe add another type of Forbidden field type, where it will behave like current Hidden field type.

At the moment, the `_etag` is not a Schema field at all, but included into the response by the `DefaultResponseFormatter`, and for the list-view only:
- https://github.com/rs/rest-layer/blob/5b33327a0fb07b058fa7f55e5c7798c1a9b07164/rest/response.go#L102

Anyway, maybe we can leave the extra `_etag` field out for now then (just include the root `_etag` as we do today), and add the compound E-tag for the list view only -- maybe even just calcualte/set it in the `DefaultFormatResponse`.

Then this is a pretty small change, and it's easy to benchmark any performance loss.

**Comment by Dragomir-Ivanov** [comment]
There is already PR for that #162

**Comment by rs** [comment]
We can have an etag that is composed of two part, one which is the etag of the root object (as stored in the db) and another which is the compound etag (like checksum of the response).

The full etag will be used by the client for caching. But when it comes to write operations, the server is the one choosing how to extract the relevant information from the etag to fulfill the If-Match condition. In this occurrence, the server would just ignore the second part of the etag that is not related to the root document, and only use the first part for comparison.

Does it make sense?

**Comment by Dragomir-Ivanov** [comment]
I think when client is writing compound items, it gets more complicated. For one client must remove all `sub-resources`, because he obviously can't PUT the whole projection. Doing so, we already enter the realm of "not that trivial", so why we don't proceed further and include the `_etag` in the root resource. 

That way compound ETag will be left solely as a fingerprint for the whole resource, and server can also check for `_etag` in the root resource body, and use it if no `If-Match` is provided. That way clients can update a resource safely, without any HTTP header modifications. Of-course all this `_etag` in resource body could be explicitly requested via custom HTTP header.

For distinct resources, say /commits/12345, we can set the ETag as the `_etag`, and probably omit the `_etag` from the resource, because it will be a duplication.

**Comment by smyrman** [comment]
> The full etag will be used by the client for caching. But when it comes to write operations, the server is the one choosing how to extract the relevant information from the etag to fulfill the If-Match condition. In this occurrence, the server would just ignore the second part of the etag that is not related to the root document, and only use the first part for comparison.

> Does it make sense?

Yes - for the single item `GET` view, this makes perfect sense. I suppose those would also be marked as strong then.

@Dragomir-Ivanov commented on the review.

**Comment by smyrman** [comment]
> I think when client is writing compound items, it gets more complicated. For one client must remove all sub-resources, because he obviously can't PUT the whole projection. Doing so, we already enter the realm of "not that trivial",

Thats a very good point actually!

**Comment by Dragomir-Ivanov** [comment]
Since we have cleared all outstanding PRs, are you willing to continue discussion on this topic. I believe it is important and this can make `rest-layer` very flexible :)

**Comment by smyrman** [comment]
It would indeed be good to continue this.  Maybe propse some plausable code for the MD5 tag prefix idea. Might require an  inteeface change.

**Comment by Dragomir-Ivanov** [comment]
Before proposing any code, lets try to agree on a way forward with this. This is what we currently have in `rest-layer`:
When returning `itemList` via `ResponseFormatter` `FormatList` we get `_etag` field unconditionally.
This query ```GET http://192.168.0.103:3001/api/clients?limit=1&fields=address```
```
[
  {
    "_etag": "W/\"a2fef762819ea9e5b27855f6a11c534c\"",
    "address": "Bulgaria, Ruse, David street 89123"
  }
]
```
So we are already polluting the item, and there is no mention we wanted `_etag` at all. Imagine we are returning the whole item, make some changes, then wanted to PUT it back we should remove `_etag` manually from the body & compose our HTTP PUT request header `If-Match` with its value. Returning `_etag` for items doesn't help with Browser caching in any way. 

Now what if, when PUT/POST an item we check HTTP Header `If-Match` or a presence of `_etag` in the body. That way we will avoid composing custom HTTP header, and will not need to remove `_etag` if we wanted conflict-free update.

What about requesting `_etag` presence at all in the GET query at all:
```GET http://192.168.0.103:3001/api/clients?limit=1&fields=address&etags=1```
That way we will present the user actual means of controlling that `_etag` pollution.

Then for composite ETag, if `_etags` are present we can MD5 just them, or if not present we can MD5 the response body. 

I will stop here for now, and discuss item with `sub-resources` after we finish with this case.

Another thought, what about if somebody wants to use `rest-layer` in Function-As-a-Service environment. All HTTP headers are lost in this case, so `_etag` in the body is very viable idea for conflict-free updates. Obviously in FaaS case all browser caches are not available.

@rs @smyrman Any feedback appreciated.

**Comment by rs** [comment]
Why a new query param and not just a projection field?

**Comment by smyrman** [comment]
Can I recommend that we splitting the issue in several parts to make it easier to solve?

## Part A

I think the first problem to solve is `GET` when a referenced objects change so that we can reliably use the E-tag to check if resources are up-to-date when embedding sub-resources. E.g.

1. `http GET 'http://192.168.0.103:3001/api/clients/<client-id>' fields=services If-Match:W/"<db-etag>"` where services is a sub-resource and/or reference that we want to expand.
2. A linked service change.
3. `http GET 'http://192.168.0.103:3001/api/clients/<client-id>' fields=services If-Match:W/"<db-etag>"` still match even though the resulting JSON is different 😢 

@rs idea could by prefixing with an MD5 could help resolve this _regardless_ if sub-resources are polluted with `_etag` fields or not.


> We can have an etag that is composed of two part, one which is the etag of the root object (as stored in the db) and another which is the compound etag (like checksum of the response).

E.g. `"<MD5-sum-of-payload>-<db etag>"`. on POST, the server will use the `<db-etag>`

For the list view, we change the implementation to let the header E-tag be a MD5 sum of the entire payload (also in case of HEAD, which means we must do more work for HEAD requests).

## Part B

> Why a new query param and not just a projection field?

We create a new issue to optionally include DB `_etags` for sub-resources and embedded refrences, either via a custom query parameter, or via some form of (special-cased) field-selection. Remember that `_etag` is not currently handled as a `schema.Field` at all.

**Comment by smyrman** [comment]
Part A is the original issue basically...

**Comment by rs** [comment]
+1

I would put the db etag first in your example.

**Comment by Dragomir-Ivanov** [comment]
Guys, I was hoping we would focus on the `itemList` case first. On one side it could give us some hints on how to resolve the `composite item` case, and on other it is somewhat easier case.

@rs In order to achieve `_etag` as a projection field, we need to include it in `item.Payload` structure. If we do that, will we include the `_etag` implicitly in the resource `Schema` or leave the user to define it.
`schema.Readonly` will not be enough, as user will surely try to pass that to subsequent PUT request.

@smyrman I fail to understand what `<MD5-sum-of-payload>-<db etag>` will give us. Does it give us better Browser cache hit rate, or just avoids us polluting the root item with `_etag` field. If latter is the case, please consider we are currently polluting all items in `itemList` case. If we are going to pollute anything, we better pollute in controlled manner(on/off) and consistently.

**Comment by smyrman** [comment]
> Browser cache hit rate, or just avoids us polluting the root item with_etag field. 

Actually neither, it helps with better cache miss rate! 

If you don't have either E-tag or no-cache headers, as you have discovered with clients, you get fake browser cache hits when you have embedded items. E.g. in our case where we have a "dashboard"-like resource with multiple "widgets" as sub-resources. If we add or remove or re-order "widgets" on the dashboard, the dashboard won't correctly refresh without deleting cache in the browser because the browser think it's up to date.

The composite E-tag with both the db-Etag and MD5 sum of the full payload would make the browser `GET` cache behave correctly while still allow for conditional `POST`, `PUT` and `PATCH` of the root resource to function, given you replace or remove embedded items.

**Comment by Dragomir-Ivanov** [comment]
Actually I am advocating in case of `composite` resource, although I wanted to avoid discussion about it right now, we would have one and only ETag for the whole resource body (`root` and `sub-resource`). Thus the HTTP Etag will be fingerprint of the entire payload. Not two sections of it. 

Lets focus on the `itemList` case. I think after resolving this, the `composite item` will be resolved naturally.

**Comment by Dragomir-Ivanov** [comment]
@smyrman I think that two-part ETag is unnecessary. Perhaps if you can give me some concrete example I can find it useful.

**Comment by rs** [comment]
With the two-part ETag you keep the opportunity to use the db etag part to perform conditional writes.

**Comment by smyrman** [comment]
For the itemList case, you _just_ need an MD5 sum of the full Json payload (no db E-tag, as there is no relevant one to use), and everything should just work, so that's a good place to start.

The two-part is only valid for the detail view for now (when ID is given in the URL), and, as @rs  mentiones, you can then do conditional writes with the same E-tag with the root resource, e.g. if I want to PATCH the "description" field of my "dashboard" only if nobody else have done so since my last read.

**Comment by Dragomir-Ivanov** [comment]
@rs I am also talking about the same thing, however I fail to see why:
`ETag: <MD5 payload>-<db etag>` is more than

`ETag: <MD5 payload>`
```
{
   "_etag": "root item itag",....
  "sub-resource1" : {
      "_etag": "sub etag",....
  },
  "sub-resource2": {
      "_etag": "sub etag",...
  }
}
```

Keep in mind, that we already have polluted `itemList` case with items `_etag`. Doing it that way, we will have consistent behavior between `itemList` and `composite item`, where one will find ETag he needs for conflict-free update in the `_etag` field in the item.

@smyrman There might be very well the case, when one extract a list of items, then wants to update one or more items from the list, without any re-fetch. So there are cases where `_etag` fields are needed. In order to calculate the composite ETag, one might use the MD5 of all `_etag` fields in the list as an optimization. As I said before, that is Strong ETag, because will consider the updates outside the current projection.

**Comment by rs** [comment]
Let's imagine you have the following projection `title,description,author{name}`:

```json
{
    "title": "A title,
    "description": "Some description",
    "author": {
        "name": "John Doe"
    }
}
```

Because author is a sub-resource, you can't just give the root (db) etag, you have to add the content of author so if the author name change you get a different etag. So as defined, you would MD5 the whole response instead of trying to get individual etags, which by itself, could be your final etag.

But now you want to PATCH the description in a CAS way. Normally you would use the Etag, but if it is an MD5 of a certain projection, we lost the ability to perform conditional writes. So using `If-Match: MD5(projection-result)` would always fail.

If we return both the root (db) etag and the MD5 of the response like `W/"<db-etag>~<md5>"`, when the client use that value with `If-Match`, we can remove everything after the ~ and use the first part to perform the conditional write.

Said otherwise, we don't break one existing feature to fix the caching.

**Comment by Dragomir-Ivanov** [comment]
I totally agree with this, however what I am saying is this. Following the **already** implemented logic in the `itemList` case we have this projection:
```
{
    "title": "A title",
    "description": "Some description",
    "_etag": "123",
    "author": {
        "name": "John Doe",
        "_etag": "345"
    }
}
```

Trying to update/PATCH the `root` resource we supply the root `_etag`: `If-Match`:[root _etag]`.

All I am proposing is to extend the current logic we have for `itemList` to `composite resources` that is all.
Along with ability to signal if we want the addional `_etag` pollution or not ( via HTTP Header, query parameter, projection field, etc. )

Browser caching will not be compromised, because the HTTP ETag header will be fingerprinf of the whole HTTP body ( root + sub-resources ).

Don't you think that is more flexible, as one can patch either to `root` resource, or any of the `sub-resources` without refetch having already their `_etag`s.

**Comment by rs** [comment]
Do you think it is practical to use the etag of a projected sub-resource to update it? In your example you don't even have it's ID and you would have to guess its path.

The item list use-case is already kind of stretched, but at least it's part of the same resource.

I think the main use-case is to use the standardized etag to perform conditional mutations without having to pick the etag into a custom field of the payload.

**Comment by smyrman** [comment]
> @smyrman There might be very well the case, when one extract a list of items, then wants to update one or more items from the list, without any re-fetch. So there are cases where _etag fields are needed.

I vave never disputed that, I am all for embedding more _etag fields, and get better control of it. I just thinks that it's a separate issue, as it doesn't help with more consistent `GET`. If we break this up I want to suggest:

1. The itemListview get a strong `MD5` from the full response body, no other changes. This could be the first PR.
2. The single item view get a strong MD5. We can not easily _know_ if we have embedded fields or not in this view at the Formatter/Sender, so here we need to either allow embedding of  the `_etag` field  as you suggest @Dragomir-Ivanov (for the root resource only to retain current functionality), _or_ we need the two-component tag that @rs suggest. Eithe would work, but the two-component one don't change the current API usage workflows.

> Do you think it is practical to use the etag of a projected sub-resource to update it? In your example you don't even have it's ID and you would have to guess its path.

The example from @Dragomir-Ivanov might not be complete, but if you embed the full sub-resource including the ID, then having the E-tag would be useful. Anyway I suggest we create a new issue / PR to address embeding of _etag in subresources, and to better control the inclusion of them in the payload, because I think this requires a bit more thought-process, not to mention the code changes. I am not against it, if I come across that way.

**Comment by Dragomir-Ivanov** [comment]
@smyrman The first post in this issue is about ETag on a composite resource, so I personally think this thread is good for discussing that matter. There is no code yet, nor PR result from this ( but I am willing to do the necessary changes when we settle on a solution ).

1. Also we already have code in `master` I did recently about ETag HTTP response header on `itemList` case. It is not strong MD5 sum from the body, but this easily can be achieved, and I don't have any objections on it (as we are all on the same page for this), and I can send PR for this right away.

2. I guess we have some misunderstanding / disagreement here. I still fail to understand what two component `_etag` will bring us. With including `_etag` in the item body we achieve consistency like using/updating the items will be the same no matter where they came from: itemList, single item ( with or without sub-resources), and sub-resource. All of them will have `_etag` in their bodies and their update procedure will be the same. With two-component ETag, we will have different update procedure for the root item, but the others 2 will rely on the embedded `_etag` field.

When supplying the `_etag` on PUT/PATCH rest-layer can use it to ensure conflict-free update ( as it came from `If-Match` ) and prevent its propagation to the Storage. 

I understand your concern about not braking the API usage, but if we are to brake something, better be before v1.0.

**Comment by Dragomir-Ivanov** [comment]
Keep in mind that in PUT/PATCH case, the user will need to set the `If-Match` HTTP header, and not the browser.

**Comment by rs** [comment]
Sure, but the HTTP spec says that you are supposed to use the value of the `Etag` header. If we don't follow the HTTP spec, we should not use the `If-Match` header because it becomes confusing.

Again, I don't like having `_etag` in the payload. If we can have it only on demand I would sleep better, and if we can make the main use-case of etag (CAS on single resource) work without this extra field it would be better.

**Comment by Dragomir-Ivanov** [comment]
@rs That was my idea the whole time. We should make this `_etag` pollution on demand ( via query parameter just like `X-Total`, or HTTP header `X-RS-Etag`, or even we can we can enable/disable it via API, `index.enablePayloadEtags(true)`. 

Now are you talking about single resource without sub-resources or with sub-resources.
1. In case without sub-resources, HTTP ETag will be the fingerprint of the resource ( i.e. the same as not returned `_etag` ).
2. In case with sub-resources, HTTP ETag I guess we can use the two-part ETag. So updating the root resource will be possible, and the whole projection can be cached by the browser. Hope this two-part ETag is not too confusing for the users. What is your proposal for the format `ETag: "[root item etag]"-"
[body md5 sum]"`, with a dash between them and quotes around each hash sum?

**Comment by smyrman** [comment]
> Now are you talking about single resource without sub-resources or with sub-resources.

In the ResponseFormatter/SendFormatter this is currently an impossible call to make, as the peojection is already applied. To keep both simpler code and a consistent user exoerience, it might be better to always go for case 2 snd use the two-part ID.

**Comment by smyrman** [comment]
> with a dash between them and quotes around each hash sum?

That breaks the spec, does it not?

**Comment by Dragomir-Ivanov** [comment]
Yes, maybe it breaks the spec. It will be ETag: "[root item tag]-[body md5 sum]".
@smyrman What do do you think _etag enable switch be?

**Comment by rs** [comment]
It does not break the spec and is not confusing to the user. The fact that it’s a composite etag is opaque to the user. The user still send the full etag, it is rest layer which choose to only use the first part for mutations.

**Comment by rs** [comment]
And I agree with @smyrman, we should start by just fixing the caching bug using composite etags. Let’s handle etag switch in a different issue.

**Comment by smyrman** [comment]
> It does not break the spec and is not confusing to the user. 

Sorry, I was only reffing to the quotes around both pars - we should have only one set of quotes.
